docs/developer/bigquery/bq-data-plane.md

```diff
@@ -0,0 +1,45 @@
+# BigQuery Data Plane
+
+This document describes the first implementation of the BigQuery data plane. 
+
+The BigQuery data plane is implemented as a pipeline by `BigQueryDataSource` and ` BigQueryDataSink` classes.
+The data source, identified by a `BigQueryDataAddress`, expects an asset defined by a query statement that, executed on a BigQuery table, returns the data to be transferred.
+The data sink, also identified by a `BigQueryDataAddress`, receives the data from the source and transfers it without changes on a BigQuery destination table.
+
+## Data source
+
+The data source supports queries with [named parameters](https://cloud.google.com/bigquery/docs/parameterized-queries). To set the values of the named parameters, the `BigQueryRequestParams` generated by `BigQueryRequestParamsProvider.provideSourceParams` (invoked from `BigQueryDataSourceFactory.createSource`) includes also the sink address passed when the transfer process is started by the consumer. The sink address must provide the values of the parameters in the query in the form:
+```
+"dataDestination": {
+    "type": "BigQueryData",
+    "project": "consumerProject",
+    "dataset": "destinationDataset",
+    "table": "destinationTable",
+    "@TYPE_parameterName": "parameterValue"
+}
+```
+The TYPE is separated by the parameter name found in the query by underscore and can take values found in:
+[StandardSQLTypeName](https://cloud.google.com/java/docs/reference/google-cloud-bigquery/latest/com.google.cloud.bigquery.StandardSQLTypeName)
+
+When `BigQueryDataSource.openPartStream` is invoked, the data source creates the job to execute the query, a `PipedOutputStream` and a connected `PipedInputStream`:
+the output stream is passed to a newly created thread, and used to write the fetched rows as JSON entries: the rows are grouped in pages (paginated results) and sent as a JSON array
+the input stream instead is passed to a single `BigQueryPart` that is immediately returned
+The `openPartStream` method returns immediately, while the created thread fetches the results and streams the results as JSON data to the sink.
```

Comment:
If I understand correctly the `outputStream` (of the single part) would receive N pages, each consisting of a JSON array of rows. So the source's outputStream receives N individual JSONArrays.

Another way of doing it would be to package individual pages into one `BigQueryPart` object _each_, and returning a stream of those parts. So there would be N parts, each containing _one page_ of M rows. This would avoid potentially materializing the entire Stream in memory.

I'm assuming there is a reasoning behind why that approach _wasn't_ taken, so I think there's value in outlining those reasons.

Reply:
The reason for having one part only in this first implementation is to simplify the design: the openPartStream can return immediately with the single part providing all the data. The alternative, using multiple parts, would require to get the total number of returned rows / pages, creating the parts each with an output, connecting each output to an input, maintaining the list in the source and then working one page at time. If this helps can be added of course.